src/python/pants/backend/python/rules/black.py

```diff
@@ -0,0 +1,148 @@
+# Copyright 2019 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from dataclasses import dataclass
+from pathlib import Path
+from typing import Optional, Tuple
+
+from pants.backend.python.rules.pex import (
+  CreatePex,
+  Pex,
+  PexInterpreterConstraints,
+  PexRequirements,
+)
+from pants.backend.python.subsystems.black import Black
+from pants.backend.python.subsystems.python_setup import PythonSetup
+from pants.backend.python.subsystems.subprocess_environment import SubprocessEncodingEnvironment
+from pants.backend.python.targets.formattable_python_target import FormattablePythonTarget
+from pants.engine.fs import Digest, DirectoriesToMerge, PathGlobs, Snapshot
+from pants.engine.isolated_process import (
+  ExecuteProcessRequest,
+  ExecuteProcessResult,
+  FallibleExecuteProcessResult,
+)
+from pants.engine.rules import optionable_rule, rule
+from pants.engine.selectors import Get
+from pants.rules.core.fmt import FmtResult
+from pants.rules.core.lint import LintResult
+
+
+@dataclass(frozen=True)
+class BlackSetup:
+  """This abstraction allows us to set up everything Black needs and, crucially, to cache this setup
+  so that it may be reused when running `fmt` vs. `lint` on the same target."""
```

Comment:
It's not clear that that is crucial... all of the individual `yield`s in `setup_black` are themselves cached, so unless there is a lot of CPU time involved in the body of that rule, attempting to cache it isn't likely to be effective.

Another thing that might make more sense would be to extract everything _except_ the sources for a target to a `@rule`: that would aggregate "build the black pex" and "capture black's config".

Both optimizations are probably premature though: any nodes that haven't been invalidated are sitting in memory, and can be grabbed cheaply.

Reply:
I think the intent with `BlackSetup` (originally `BlackInput`) is more than anything to deduplicate the `fmt` and `lint` rule bodies. The sole differences between the two rules are:

- pass `--check` to Black if linting
- `FallibleExecuteProcessResult` vs `ExecuteProcessResult`
- converting the FEPR or EPR into `FmtResult` vs. `LintResult`

This refactored implementation avoids any duplication to only have those 3 differences contained in the body o` fmt` and `lint`.

--

Compare this to `python_test_runner`, where we instead have one really long body.

--

I can update this docstring to explain that really the motivation is deduplication, not necessarily caching.

Reply:
Yeah, I agree that this is useful for encapsulation and code factoring, but not super crucial :)

Reply:
Sounds good, thanks.